src/common/src/types/macros.rs

```diff
@@ -0,0 +1,301 @@
+macro_rules! for_all_array_variants {
+    ($macro:ident $(, $x:tt)*) => {
+        for_all_variants! { project_array_variants, $macro, [ $($x, )* ] }
+    };
+}
+#[macro_export]
+macro_rules! project_array_variants {
+    ($macro:ident, [ $($x:tt, )* ], $( { $data_type:ident, $variant_name:ident, $suffix_name:ident, $scalar:ty, $scalar_ref:ty, $array:ty, $builder:ty } ),*) => {
+        $macro! {
+            $($x, )*
+            $( { $variant_name, $suffix_name, $array, $builder } ),*
+        }
+    };
+}
+
+/// The projected version of `for_all_variants` for handling mapping of data types and array types.
+///
+/// Arguments are `$data_type`, `$variant_name`.
+#[macro_export(local_inner_macros)]
+macro_rules! for_all_type_pairs {
+    ($macro:ident $(, $x:tt)*) => {
+        for_all_variants! { project_type_pairs, $macro, [ $($x, )* ] }
+    };
+}
+#[macro_export]
+macro_rules! project_type_pairs {
+    ($macro:ident, [ $($x:tt, )* ], $( { $data_type:ident, $variant_name:ident, $suffix_name:ident, $scalar:ty, $scalar_ref:ty, $array:ty, $builder:ty } ),*) => {
+        $macro! {
+            $($x, )*
+            $( { $data_type, $variant_name } ),*
+        }
+    };
+}
+
+/// Helper macro for expanding type aliases and constants. Internally used by `dispatch_` macros.
+#[macro_export]
+macro_rules! do_expand_alias {
+    ($array:ty, $variant_name:ident, (
+        $(Array, $array_alias:ident,)?
+        $(ArrayBuilder, $array_builder_alias:ident,)?
+        $(Scalar, $scalar_alias:ident,)?
+        $(ScalarRef, $scalar_ref_alias:ident,)?
+        $(VARIANT_NAME, $variant_name_alias:ident,)?
+    )) => {
+        $(type $array_alias = $array;)?
+        $(type $array_builder_alias = <$array as $crate::array::Array>::Builder;)?
+        $(type $scalar_alias = <$array as $crate::array::Array>::OwnedItem;)?
+        $(type $scalar_ref_alias<'scalar> = <$array as $crate::array::Array>::RefItem<'scalar>;)?
+        $(const $variant_name_alias: &'static str = stringify!($variant_name);)?
+    };
+}
+
+/// Helper macro for generating dispatching code. Internally used by `dispatch_xx_variants` macros.
+#[macro_export(local_inner_macros)]
+macro_rules! do_dispatch_variants {
+    // Use `tt` for `$alias` as a workaround of nested repetition.
+    ($impl:expr, $type:ident, $inner:pat, [$alias:tt], $body:tt, $( { $data_type:ident, $variant_name:ident, $suffix_name:ident, $scalar:ty, $scalar_ref:ty, $array:ty, $builder:ty } ),*) => {
+        match $impl {
+            $( $type::$variant_name($inner) => {
+                do_expand_alias!($array, $variant_name, $alias);
+                #[allow(unused_braces)]
+                $body
+            }, )*
+        }
+    };
+}
+
+/// Dispatch the code block to all variants of `ArrayImpl`.
+///
+/// # Usage
+///
+/// The basic usage to access the inner concrete `impl Array` value is:
+///
+/// ```ignore
+/// fn do_stuff<A: Array>(array: &A) { .. }
+///
+/// fn do_stuff_impl(array_impl: &ArrayImpl) {
```

Comment:
What about `do_stuff_dispatch`? `do_stuff_impl` is somewhat ambiguous, sometime we use `foo_impl` to represent internal implementation of a public function `foo`.

Reply:
Okay